File: 1-100/67-add-binary.cpp
string addBinary(string a, string b) {
    int m = a.size(), n = b.size();
    reverse(a.begin(), a.end());
    reverse(b.begin(), b.end());
    string ans;
    int bit_a, bit_b, sum, carry = 0;
    for (int i = 0; i < max(m, n); ++i) {
        bit_a = (i < m ? a[i] - '0' : 0);
        bit_b = (i < n ? b[i] - '0' : 0);
        sum = (bit_a + bit_b + carry) % 2;
        carry = (bit_a + bit_b + carry) / 2;
        ans += to_string(sum);
    }
    if (carry != 0) {
        ans += "1";
    }
    reverse(ans.begin(), ans.end());
    return ans;
}

File: python version/67-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        a = a[::-1]
        b = b[::-1]
        m = len(a)
        n = len(b)
        carry = 0
        ans = ""
        for i in range(max(m, n)):
            a_bit = int(a[i]) if i < m else 0
            b_bit = int(b[i]) if i < n else 0
            bit_sum = a_bit + b_bit + carry
            carry = bit_sum // 2
            ans += str(bit_sum % 2)
        
        if carry:
            ans += "1"
        
        return ans[::-1]